kvcoord: support BATCH_RESPONSE scan format with buffered writes

yuzefovich opened the pull request:
This PR contains several commits that add support for BATCH_RESPONSE scan format of Scan and ReverseScan requests to the buffered writes interceptor. It refactors the existing logic for KEY_VALUES format to make it more extensible, and then maintains the same "hierarchy" of `[][]byte` slices in the response. It also fixes `NumKeys` and `NumBytes` fields of the merged response. See each commit for more details.

Fixes: #141743

yuzefovich committed:
kvclient: inline some helper methods in txnWriteBuffer helpers

Given that we only have two scan variants (Scan and ReverseScan) and
will only support two scan formats for now (KEY_VALUES and
BATCH_RESPONSE), it seems cleaner and less verbose to use `if` blocks
rather than `switch` statements. Thus, this commit inlines some of the
helper methods and gets rid of `switch`es. I think it'll be easier to
follow the logic for BATCH_RESPONSE scan format added in the following
commit.

The only somewhat meaningful change is the fact that we now check the
scan format first as opposed to checking whether we have a Scan or
a ReverseScan (this will make logic for BATCH_RESPONSE cleaner)

Release note: None

yuzefovich committed:
kvcoord: refactor txnWriteBuffer helpers

This commit refactors a couple of helpers that we use when merging Scan
and ReverseScan responses with the buffered values. This is done in
preparation for supporting BATCH_RESPONSE scan format. In particular:
- it introduces a helper struct that is used on the first iteration to
calculate the size of the responses
- it "disassembles" the responses into its contents (i.e. `Rows` field).
This also allows us to remove some code duplication between Scans and
ReverseScans.

It also exposes the need to adjust NumKeys and NumBytes fields of the
response. Addressing that is left as a TODO.

Release note: None

yuzefovich committed:
kvcoord: support BATCH_RESPONSE scan format with buffered writes

This commit extends the txnWriteBuffer helper structs to support
BATCH_RESPONSE scan format of Scans and ReverseScans. In this format
`BatchResponses` field contains `[][]byte` slices where a single
`[]byte` slice can have multiple KVs in the "engine" format. (Each
`[]byte` is growing in size exponentially, until a hard limit, for large
scans.) The contribution of this commit is two-fold:
- the size helper struct is responsible for calculating the precise
capacity that each `[]byte` is needed in the merged response
- during the main iteration, when accepting the KV from the buffer we
need to write it in the same "engine" format as the KV server did.

For simplicity, this commit keeps the hierarchy of slices in
`BatchResponses` unchanged, i.e. we inject all buffered KVs into the
"current position" of `BatchResponses[i]` from the server response. This
means that in the extreme case when the server response is empty, all
buffered KVs will be included in a single slice. This could be
suboptimial from memory GC perspective, but adding a smarter heuristic
is left as a TODO.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142439)
<!-- Reviewable:end -->

yuzefovich commented:
@arulajmani if you have time, this is mostly ready for a look, the last 4 commits belong to this PR. In the last commit I still have a few things to do (add tests and figure out whether / how we can avoid a dependency on `pkg/storage`), but the general structure should remain the same.

arulajmani reviewed (commented):
Great work! The overall structure here looks good

> Reviewed 14 of 14 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-:-OLEzMP8Aa-4iWRl9L69:b-b47hq6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLEt_c3B_0ILTUraMAU:-OLEt_c3B_0ILTUraMAV:b-7qmukd):*
The disallowed_imports_test is going to be mad at you for this.
___
*[`pkg/storage/enginepb/decode.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLErc2ID1bEL7ca0qSi:-OLErc2ID1bEL7ca0qSj:bbldrql) ([raw file](https://github.com/cockroachdb/cockroach/blob/598ee5e1c7277bc184e43762aa0fbcf70ff921ee/pkg/storage/enginepb/decode.go#L31)):*
> ```Go
> 	}
> 	// Last byte is the version length + 1 when there is a version,
> 	// else it is 0.
> ```

nit: this commentary could be pulled out into its own separate commit.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1019 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLEyIEx-AsiPqEMUciQ:-OLEyIEx-AsiPqEMUciR:b-lnfdqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/10e884db290a0b3d5aedd11595038cd6dec7d1cc/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1019)):*
> ```Go
> 	respEmpty bool
> 
> 	// rows is the Rows field of the corresponding response.
> ```

nit: may be worth clarifying that this is set iff scanFormat == KEY_VALUES; similar comment for batchResponses.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1079 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLEvWyE3hFQWfQkKSjr:-OLEvWyE3hFQWfQkKSjs:bbgside) ([raw file](https://github.com/cockroachdb/cockroach/blob/10e884db290a0b3d5aedd11595038cd6dec7d1cc/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1079)):*
> ```Go
> // peekKey returns the key at the current iterator position.
> //
> // peekKey will only be called if the iteration is in valid state (i.e. valid()
> ```

nit: "should only be called"? (below as well).
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLExJUhCHmH9LBzkl1V:-OLExJUhCHmH9LBzkl1W:b-abnvqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/10e884db290a0b3d5aedd11595038cd6dec7d1cc/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1275)):*
> ```Go
> 	if !it.valid() {
> 		// If respIter is exhausted, all the remaining buffered KVs go into the
> 		// very last slice.
> ```

I don't have a reason to opine either way, but curious why we're appending to the very last slice instead of adding a new one?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
This PR is still missing tests, but otherwise is RFAL. (I only adapted a single sanity test so far since `TestTxnBufferedWritesOverlappingScan` utilizes API that hard-codes `KEY_VALUES` format.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-:-OLHGDqm3UjBjez2XnCk:byr0bav)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`-- commits` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLEt_c3B_0ILTUraMAU:-OLHF5TTAs2xjDT1mzMM:b-os9iha):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The disallowed_imports_test is going to be mad at you for this.
</blockquote></details>

Refactored by introducing a new `mvccencoding` package.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142439#-OLExJUhCHmH9LBzkl1V:-OLHEsrS79rdv9jlJIfz:b-dzt3er) ([raw file](https://github.com/cockroachdb/cockroach/blob/10e884db290a0b3d5aedd11595038cd6dec7d1cc/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1275)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't have a reason to opine either way, but curious why we're appending to the very last slice instead of adding a new one?
</blockquote></details>

I thought it would be easier to handle it this way. I refactored the code to include a new "spill-over" slice which overall actually makes the code a bit simpler.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
kvcoord: correctly set NumKeys and NumBytes on merged scan responses

This commit utilizes the respSizeHelper struct to calculate the final
values for `NumKeys` and `NumBytes` fields of merged Scan and ReverseScan
responses when they are produced by the txnWriteBuffer. The existing
test has been extended to verify these values which required some minor
adjustments to correctly simulate already committed expected values
(they have the timestamp set).

Release note: None

yuzefovich commented:
Alright, this is RFAL (I added a unit test as well as fixed `NumKeys` / `NumBytes` in a new commit).

herkolategan commented:
We're still busy testing Microbenchmarks CI, and the posts are hidden for now (in the future it would have triggered on this PR): https://github.com/cockroachdb/cockroach/actions/runs/13915057320?pr=142439
We now do 3x runs on CI that all have to be positive (regressed) with a very low p-value.

<strong>🔴 Sysbench</strong> [KV, 1node, local, oltp_write_only]

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| 🔴 **sec/op** | 896.1µ ±0% | 903.0µ ±0% | +0.76% | p=0.000 n=60    |
| 🔴 **allocs/op** | 1.356k ±0% | 1.357k ±0% | +0.07% | p=0.000 n=60    |

Take it with a grain of salt, as we're still testing (any feedback welcome).
I was able to reproduce it on my `gceworker` as well.

yuzefovich commented:
Hm, @herkolategan I'd be very surprised if this PR introduces any regression because all of the code being modified is currently disabled (except in unit tests). I kicked off a run on the gceworker and didn't observe a regression:
```
yahor_cockroachlabs_com@gceworker-yuzefovich:~/go/src/github.com/cockroachdb/cockroach$ benchdiff --run=^BenchmarkSysbench/KV/1node_local/oltp_write_only$ --old=96298a --new=69b8e4d ./pkg/sql/tests -b
old:  96298a4 Merge #142768 #143009
new:  69b8e4d kvcoord: correctly set NumKeys and NumBytes on mer
args: benchdiff "--run=^BenchmarkSysbench/KV/1node_local/oltp_write_only$" "--old=96298a" "--new=69b8e4d" "./pkg/sql/tests" "-b"

name                                        old time/op    new time/op    delta
Sysbench/KV/1node_local/oltp_write_only-24    1.17ms ± 1%    1.15ms ± 3%   ~     (p=0.237 n=8+10)

name                                        old errs/op    new errs/op    delta
Sysbench/KV/1node_local/oltp_write_only-24      0.00           0.00        ~     (all equal)

name                                        old alloc/op   new alloc/op   delta
Sysbench/KV/1node_local/oltp_write_only-24     263kB ± 2%     264kB ± 1%   ~     (p=0.739 n=10+10)

name                                        old allocs/op  new allocs/op  delta
Sysbench/KV/1node_local/oltp_write_only-24     1.38k ± 0%     1.38k ± 0%   ~     (p=0.214 n=10+9)
```

herkolategan commented:
> Hm, @herkolategan I'd be very surprised if this PR introduces any regression because all of the code being modified is currently disabled (except in unit tests). I kicked off a run on the gceworker and didn't observe a regression:

Thanks for checking! You're right, there's no action to be taken here.

Strangely enough this reliably reproduce on a `c2-standard-16` machine (similar to what CI runs on `c4-standard-8`); **But** it does not reproduce on the `n2-custom-24-32768` machines most `gceworkers` are on.

We also inspected the cpu profile diffs, and nothing in this PR coincides with where the slowdown originates from.

yuzefovich commented:
TFTRs!

bors r+
